blob/blob.go

```diff
@@ -0,0 +1,55 @@
+package blob
+
+import (
+	"github.com/celestiaorg/celestia-app/x/blob/types"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+// Commitment is a Merkle Root of the subtree built from shares of the Blob.
+// It is computed by splitting the blob into shares and building the Merkle subtree to be included after Submit.
+type Commitment []byte
+
+// Proof is a collection of nmt.Proofs that verifies the inclusion of the data.
+type Proof []nmt.Proof
+
+// Blob represents any application-specific binary data that anyone can submit to Celestia.
+type Blob struct {
+	*types.Blob
+
+	commitment Commitment
```

Comment:
Later when users have to send blobs over the API (so don't have access to the constructor), they shouldn't need to include the commitment in the request - so `Commitment() Commitment` should probably check if nil, and call `CreateCommitment` if it doesn't exist

Reply:
I would avoid using Mutex just for one case.

Reply:
hmmm. my main wish is about it being optional in the json request, which would leave the commitment field empty when sending it over the wire. This means the node that receives it, will receive it without the commitment. In this case, it should be computed on the fly. I think for this reason its fine if we have a nil check in there - in 99% of cases, users will have used the constructor anyways so it will already be filled.

Your point about the mutex is good - I would leave out the mutex, and leave a comment in the godoc maybe? WDYT?

Reply:
So are you convincing me to commit a crime? 😆
Mutex department will arrest me, lol

Reply:
cc @Wondertan thoughts?

Reply:
I don't get why you think introducing mutex here is problematic. I am all for lazy computing of computing. It simplifies API with the cost of one mutex. We don't have to introduce it, though and can comment that it's not concurrency friendly, but it's nice to have it by default

Reply:
By simplifying, I mean it's unnecessary to use a constructor, and the user can simply use struct literal

Reply:
> I don't get why you think introducing mutex here is problematic. I am all for lazy computing of computing. It simplifies API with the cost of one mutex. We don't have to introduce it, though and can comment that it's not concurrency friendly, but it's nice to have it by default

I did not say it is problematic. Just wanted to avoid unnecessary mutex(if construct commitment in the c-tor)

Reply:
This mutex will be needed only once - when commitment is empty and we have to construct it. For other cases, it will be read-only operations.

Reply:
> By simplifying, I mean it's unnecessary to use a constructor, and the user can simply use struct literal

wdym? using struct literal to pass there what? Currently, we have to call `types.NewBlob(nID, data)` as it also checks the validity of the namespace inside + the length of the blob. Moving towards the structure literal, we have to ensure the blob is correct each time we will interact with it(meaning receive it from the user) and this is a huge overhead in comparison with a c-tor that will check the validity of data before creating a blob.